File: front/src/types/source.ts
import { z } from 'zod';

export enum SourceProvider {
  YOUTUBE = 'youtube',
}

export enum SourceStatus {
  PENDING = 'pending',
  PROCESSING = 'processing',
  DONE = 'done',
  ERROR = 'error',
}

export const sourceProviderSchema = z.nativeEnum(SourceProvider);
export const sourceStatusSchema = z.nativeEnum(SourceStatus);

export const sourceBaseSchema = z.object({
  provider: sourceProviderSchema,
  contentId: z.string(),
});

export const sourceInSchema = sourceBaseSchema.extend({
  metadata: z.record(z.any()).optional().default({}),
});

export const sourceSchema = sourceBaseSchema.extend({
  status: sourceStatusSchema.default(SourceStatus.PENDING),
  metadata: z.record(z.any()),
  createdAt: z.date().default(() => new Date()),
});

export type SourceBase = z.infer<typeof sourceBaseSchema>;
export type SourceIn = z.infer<typeof sourceInSchema>;
export type Source = z.infer<typeof sourceSchema>;

File: back/models/source.py
from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from models.elasticsearch import BaseIndexItem


class SourceProvider(str, Enum):
    YOUTUBE = "youtube"


class SourceStatus(str, Enum):
    PENDING = "pending"
    PROCESSING = "processing"
    DONE = "done"
    ERROR = "error"


class SourceBase(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    provider: SourceProvider
    content_id: str = Field(alias="contentId")


class SourceIn(SourceBase):
    metadata: Optional[dict] = Field(default_factory=dict)


class SourceInModifier(SourceBase):
    status: SourceStatus = SourceStatus.PENDING
    metadata: dict = Field(default_factory=dict)
    created_at: datetime = Field(default_factory=datetime.now)


class Source(SourceInModifier, BaseIndexItem):
    pass
